### ISSUE TITLE
Reset Monitor Results in Duplicated Bindings

### ISSUE THREAD

**Issue by BlueFinBima** [HeliosVirtualCockpit/Helios#490]
The profile attached is a simple testcase and the profile can be made to migrate correctly or not.  The profile has two identical controls, each with 4 bindings.  The first control is directly on the monitor and the second is on a panel.  The four bindings directly on the monitor migrate correctly to the new monitor, however the bindings on the panel are duplicated twice (ie each binding appears three times) when the migration fails.  The number of resulting endpoints in the message below does not match the endpoints I have in the migrated profile (ie 32). 

![image](https://user-images.githubusercontent.com/18526232/143293650-6db5ef32-7c32-4fe3-8acb-cd02f5ea78cc.png)

The reset monitors I preform is:

![image](https://user-images.githubusercontent.com/18526232/143299620-6687ed8d-7cac-492a-b60f-97a0c4d0dabc.png)

Empirically, what determines whether the reset monitors works correctly or not is whether there is a tab open for the panel in the centre workspace.   
![image](https://user-images.githubusercontent.com/18526232/143299966-0ef98c85-5ab0-46d1-9957-b337d27a7862.png)

If it is open, I get the extra bindings, and if the tab for the panel is closed so that it is not on the workspace, the **Reset Monitors** works.  

````
<?xml version="1.0" encoding="utf-8"?>
<HeliosProfile>
  <Version>3</Version>
  <Monitors>
    <Monitor>
      <Location>0,
[DuplicateBindingTestcase.zip](https://github.com/HeliosVirtualCockpit/Helios/files/7598261/DuplicateBindingTestcase.zip)
0</Location>
      <Size>1920,1080</Size>
      <Hidden>False</Hidden>
      <Orientation>DMDO_DEFAULT</Orientation>
      <AlwaysOnTop>True</AlwaysOnTop>
      <Background />
      <Children>
        <Control TypeIdentifier="Helios.Panel" Name="Panel A" SnapTarget="True" Locked="False">
          <Location>50,0</Location>
          <Size>300,421</Size>
          <Hidden>False</Hidden>
          <Background />
          <Interaction>
            <AllowInteractionFull>False</AllowInteractionFull>
            <AllowInteractionNone>True</AllowInteractionNone>
            <AllowInteractionLegacy>False</AllowInteractionLegacy>
          </Interaction>
          <Children>
            <Control TypeIdentifier="Helios.AV8B.ThreatIndicatorPanel" Name="Threat Indicator Panel" SnapTarget="True" Locked="False">
              <Location>0,0</Location>
              <Size>251,361</Size>
              <Hidden>False</Hidden>
              <Children />
            </Control>
          </Children>
        </Control>
        <Control TypeIdentifier="Helios.AV8B.ThreatIndicatorPanel" Name="Threat Indicator Panel" SnapTarget="True" Locked="False">
          <Location>0,0</Location>
          <Size>251,361</Size>
          <Hidden>False</Hidden>
          <Children />
        </Control>
      </Children>
    </Monitor>
  </Monitors>
  <Interfaces>
    <Interface TypeIdentifier="Helios.Base.ProfileInterface" Name="Profile" />
    <Interface TypeIdentifier="Helios.Base.Keyboard" Name="Keyboard">
      <KeyDelay>30</KeyDelay>
    </Interface>
    <Interface TypeIdentifier="Helios.AV8B" Name="DCS AV-8B" />
    <Interface TypeIdentifier="Patching.DCS.AdditionalViewports" Name="DCS Additional Viewports" />
    <Interface TypeIdentifier="Patching.DCS.MonitorSetup" Name="DCS Monitor Setup" />
  </Interfaces>
  <Bindings>
    <Binding BypassCascadingTriggers="True">
      <Trigger Source="Interface;;Helios.AV8B;DCS AV-8B" Name="Threat Indicators.SAM.changed" />
      <Action Target="Visual;Monitor 1.Panel A.Threat Indicator Panel.Threat Indicator Panel_SAM;Helios.Base.Indicator;Threat Indicator Panel_SAM" Name="Threat Indicator Panel_SAM.set.indicator" />
      <TriggerValue />
    </Binding>
    <Binding BypassCascadingTriggers="True">
      <Trigger Source="Interface;;Helios.AV8B;DCS AV-8B" Name="Threat Indicators.CW.changed" />
      <Action Target="Visual;Monitor 1.Panel A.Threat Indicator Panel.Threat Indicator Panel_CW;Helios.Base.Indicator;Threat Indicator Panel_CW" Name="Threat Indicator Panel_CW.set.indicator" />
      <TriggerValue />
    </Binding>
    <Binding BypassCascadingTriggers="True">
      <Trigger Source="Interface;;Helios.AV8B;DCS AV-8B" Name="Threat Indicators.AI.changed" />
      <Action Target="Visual;Monitor 1.Panel A.Threat Indicator Panel.Threat Indicator Panel_AI;Helios.Base.Indicator;Threat Indicator Panel_AI" Name="Threat Indicator Panel_AI.set.indicator" />
      <TriggerValue />
    </Binding>
    <Binding BypassCascadingTriggers="True">
      <Trigger Source="Interface;;Helios.AV8B;DCS AV-8B" Name="Threat Indicators.AAA.changed" />
      <Action Target="Visual;Monitor 1.Panel A.Threat Indicator Panel.Threat Indicator Panel_AAA;Helios.Base.Indicator;Threat Indicator Panel_AAA" Name="Threat Indicator Panel_AAA.set.indicator" />
      <TriggerValue />
    </Binding>
    <Binding BypassCascadingTriggers="True">
      <Trigger Source="Interface;;Helios.AV8B;DCS AV-8B" Name="Threat Indicators.SAM.changed" />
      <Action Target="Visual;Monitor 1.Threat Indicator Panel.Threat Indicator Panel_SAM;Helios.Base.Indicator;Threat Indicator Panel_SAM" Name="Threat Indicator Panel_SAM.set.indicator" />
      <TriggerValue />
    </Binding>
    <Binding BypassCascadingTriggers="True">
      <Trigger Source="Interface;;Helios.AV8B;DCS AV-8B" Name="Threat Indicators.CW.changed" />
      <Action Target="Visual;Monitor 1.Threat Indicator Panel.Threat Indicator Panel_CW;Helios.Base.Indicator;Threat Indicator Panel_CW" Name="Threat Indicator Panel_CW.set.indicator" />
      <TriggerValue />
    </Binding>
    <Binding BypassCascadingTriggers="True">
      <Trigger Source="Interface;;Helios.AV8B;DCS AV-8B" Name="Threat Indicators.AI.changed" />
      <Action Target="Visual;Monitor 1.Threat Indicator Panel.Threat Indicator Panel_AI;Helios.Base.Indicator;Threat Indicator Panel_AI" Name="Threat Indicator Panel_AI.set.indicator" />
      <TriggerValue />
    </Binding>
    <Binding BypassCascadingTriggers="True">
      <Trigger Source="Interface;;Helios.AV8B;DCS AV-8B" Name="Threat Indicators.AAA.changed" />
      <Action Target="Visual;Monitor 1.Threat Indicator Panel.Threat Indicator Panel_AAA;Helios.Base.Indicator;Threat Indicator Panel_AAA" Name="Threat Indicator Panel_AAA.set.indicator" />
      <TriggerValue />
    </Binding>
  </Bindings>
</HeliosProfile>
````
[DuplicateBindingTestcase.zip](https://github.com/HeliosVirtualCockpit/Helios/files/7598266/DuplicateBindingTestcase.zip)

**Comment by BlueFinBima** [comment]
I don't have time to look at this, but I think it might be useful if someone else could confirm this recreation scenario and findings.

**Comment by Todd1215** [comment]
I can confirm that the error presents itself it the panel has been opened and is opened before performing the monitor reset. I got the same error message. Conversely I did not get the error if you leave the panel closed and do the monitor reset.
![image](https://user-images.githubusercontent.com/6605805/143308747-08ec2afe-0777-43f3-ac67-0ead9563ef10.png)

**Comment by MedievalBunny** [comment]
I wonder if this is related to a user report that performing a monitor reset is causing bindings to disappear as well? I've been meaning to see if I can reproduce the issue and haven't had the time yet.

https://forum.dcs.world/topic/286561-apache-profile-using-helios-for-a-second-monitor/?do=findComment&comment=4959795

**Comment by BlueFinBima** [comment]
## Investigation Update

The problem lies in code already flagged by @derammo as needing a rewrite - specifically the reset 
monitors / cut & paste (no surprise).  The problem seems to be that when a panel is opened in Profile Explorer, the VisualContainerProperty changes and this results in `VisualContainer.DesignMode = true;` (line 243 of HeliosVisualContainerEditor).  This then results in a re-bind happening in CompositeVisual's OnProfileChanged method because we drop through the "bail-out" test ` if (!DesignMode)  return;`.  If the Panel is not opened in PE, then DesignMode on the panel visual remains false, and CompositeVisual never creates brand new additional bindings.

Interestingly, the situation with Cut&Paste for a CompositeVisual seems to be worse than Reset Monitors, because irrespective of the involvement of a panel, the CompositeVisual will get an additional complete set of new bindings when it is pasted.

Increasing the "bail-out" test in CompositeVisual to also check the DesignMode on the oldProfile seems to work for Reset Monitors but that does not help with Paste because the OnProfileChanged profile is null.

I have neither skills nor the courage to completely re-design and re-write this whole area (before derammo suggests it).

**Comment by BlueFinBima** [comment]
@MedievalBunny Alex, many thanks for the pointer above.  Hopefully we can reduce the instances where there are duplicate bindings.

**Comment by BlueFinBima** [comment]
Unfortunately with 1.6.5500.0, I shipped some CompositeVisual gauges which did autobinding where the trigger and action were both visuals.  These new autobindings are causing duplicate binds on reset monitors so the previous commit does not cater for this situation because they are not bound in CompositeVisual.cs

**Comment by BlueFinBima** [comment]
The code for the self bindings wasn't quite right in that it created the self bind in the Visual with the result that the the bind would happen irrespective of whether it was in DesignMode.  The code for the self bind has now been altered such that the visual creates a DefaultSelfBind, in the same manner as DefaultIn/OutputBindings, and we now leave it up to the CompositeVisual OnProfileChanged to determine if the actual bind is needed.  This is the last part (I think) of this issue.  The fix for this is in commit d72117ce8f18441154338f6fa6d489bf5dd0ba0d

**Comment by BlueFinBima** [comment]
Delivered in 1.6.5501.0
